hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/ClientSmartListenerService.java

```diff
@@ -74,7 +77,7 @@ public String call() {
                         invoke(registrationKey, connection);
                     }
                 } catch (Exception e) {
-                    deregisterListener(userRegistrationId);
```

Comment:
`The change introduced a behaviour difference` Yes and no. 
Yes, it is a behaviour change because otherwise it was always deadlock when this method is called. 
No, it is not a behaviour change because we were already waiting for future inside `deregisterListener` . This means intention was to wait for cleanup to finish in case of failure. 

`Is this the behaviour we want?`
I think it is. We don't want to have a listener getting events after registerListener saying it is failed to register.